### ISSUE TITLE
How to support custom layer parameters, e.g. dropout rates?

### ISSUE THREAD

**Issue by alexjc** [aigamedev/scikit-neuralnetwork#18]
This is the plan to handle all remaining open Issues in one go :-)

My instinct would be to extend the `layers=[]` input list to include additional named parameters. Maxout layer types currently require both number of units and the number of piecewise linear segments of the activation function. This could be extended as follows:

````
nn = MultiLayerPerceptronRegressor(
    layers=[
        L("Convolution", kernel=(3,3)),
        L("Pooling", size=(2,2)),
        L("Rectifier", 256, dropout=0.5),
        L("Maxout", 128, segments=2, dropout=0.25)])
````

The down side is the comment mentioned here, it may be harder to integrate into `sklearn` grid-search support:
http://danielnouri.org/notes/2014/12/17/using-convolutional-neural-nets-to-detect-facial-keypoints-tutorial/#comment-1756803150

I think we can also use the elegant initializer list then use custom attributes to override them, which may be the best compromise?

What do you think? @ssamot and anyone else lurking!

**Comment by alexjc** [comment]
Different syntax using multiple wrapper objects of different types, which could be put into a separate module called `sknn.layers` for example.

````
nn = MultiLayerPerceptronClassifier(
    layers=[
        Convolution(channels=8, kernel=(3,3)),
        Pooling(size=(2,2)),
        Rectifier(256, dropout=0.5),
        Maxout(128, segments=2, dropout=0.25)])
````

**Comment by ssamot** [comment]
My only worry is that having different objects for different layer types/parameters is thatt you have to import all these different things

Possibly also you might have stuff like 

````
import mlp.layers as l

nn = MultiLayerPerceptronClassifier(
    layers=[
        l.Convolution(l.channels=8, l.kernel=(3,3)),
        l.Pooling(l.size=(2,2)),
        l.Rectifier(l.units=256, l.dropout=0.5),
        l.Maxout(l.units=128, l.segments=2, l.dropout=0.25)])

````

**Comment by alexjc** [comment]
It's up to individual style.  I think we should keep the `Layers` base class.  Specific sub-classes could help with error checking.  And for import hassle, we could do:

```
from mlp.layers import *
```

**Comment by ssamot** [comment]
That's polluting the namespace me thinks :P. 

Also dropout is a cost function. Things like adding a dropout parameter to one layer (or a weight decay parameter for that matter) will influence the cost function.

**Comment by alexjc** [comment]
If you don't want to pollute the namespace you can just:
````
from mlp import layers
````

or even...

````
from mlp import layers as L
````

It's a standard Python style issue, how users get around it is up-to them.

**Comment by alexjc** [comment]
Yes, all dropout parameters are combined together for ``pylearn2``, but since they support layer-specific dropout rates and since it's commonly used in image recognition, why not support it?

**Comment by alexjc** [comment]
Also, I propose only supporting the object-based `Layer()` syntax and no longer tuples... 1, 2, 3. :+1:

**Comment by ssamot** [comment]
Doesn't make sense to support both +1
